### ISSUE TITLE
[BUG]: Using the `Microsoft.Extensions.Http.Resilience` NuGet package assumes `IRedactorService`.

### ISSUE THREAD

**Issue by IEvangelist** [dotnet/extensions#4575]
### Description

When attempting to use the `Microsoft.Extensions.Http.Resilience` package (version 8.0.0-rtm.23474.4), after calling various `.Add*Handler` methods and requesting the service that was dependent on the `HttpClient` and configured for resiliency, it throws an `InvalidOperationException`.

> No service of tyep 'Microsoft.Extensions.Compliance.Redaction.IRedactorProvider' has been registerd.

![image](https://github.com/dotnet/extensions/assets/7679720/b8ec44b1-8614-4551-a910-9ba6e3347494)

I would expect that if the `Microsoft.Extensions.Http.Resilience` library was to expose a means of registering my resilience handlers, and that if it internal is relying on an implementation of `IRedactorService` it should be responsible for taking a transient dependency on the `Microsoft.Extensions.Compliance.Readtion` library and call the appropriate DI registration methods to not surface this error. In other words, the consumer shouldn't need to know about `IRedactorService`, nor should it need to provide an implementation.

### Reproduction Steps

Here is my project file:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <Nullable>enable</Nullable>
    <ImplicitUsings>true</ImplicitUsings>
    <OutputType>Exe</OutputType>
    <LangVersion>preview</LangVersion>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Extensions.Hosting" Version="8.0.0-rc.2.23464.16" />
    <PackageReference Include="Microsoft.Extensions.Http" Version="8.0.0-rc.2.23464.16" />
    <PackageReference Include="Microsoft.Extensions.Http.Resilience" Version="8.0.0-rtm.23474.4" />
  </ItemGroup>

</Project>
```

And here's my _appsettings.json_:

```json
{
    "RetryOptions": {
        "Retry": {
            "ShouldRetryAfterHeader": false,
            "Backoff": "Linear",
            "UseJitter": false,
            "MaxRetryAttempts": 3
        }
    }
}
```

And here's the corresponding _Program.cs_:

```cs
using System.Net;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Http.Resilience;
using Microsoft.Extensions.Options;
using Polly;
using Polly.Retry;

var builder = Host.CreateApplicationBuilder(args);

var httpClientBuilder = builder.Services.AddHttpClient<ExampleClient>(
    configureClient: static client =>
    {
        client.BaseAddress = new("https://jsonplaceholder.typicode.com");
    });

httpClientBuilder.AddStandardResilienceHandler(options =>
{
    options.RetryOptions.BackoffType = DelayBackoffType.Linear;
});

httpClientBuilder.AddResilienceHandler("CustomPipeline", static builder =>
{
    // See: https://www.pollydocs.org/strategies/retry.html
    builder.AddRetry(new HttpRetryStrategyOptions
    {
        // Customize and configure the retry logic.
        BackoffType = DelayBackoffType.Exponential,
        MaxRetryAttempts = 5,
        UseJitter = true
    });

    // See: https://www.pollydocs.org/strategies/circuit-breaker.html
    builder.AddCircuitBreaker(new HttpCircuitBreakerStrategyOptions
    {
        // Customize and configure the circuit breaker logic.
        SamplingDuration = TimeSpan.FromSeconds(10),
        FailureRatio = 0.2,
        MinimumThroughput = 3,
        ShouldHandle = static args =>
        {
            return ValueTask.FromResult(args is
            {
                Outcome.Result.StatusCode:
                    HttpStatusCode.RequestTimeout or HttpStatusCode.TooManyRequests
            });
        }
    });

    // See: https://www.pollydocs.org/strategies/timeout.html
    builder.AddTimeout(TimeSpan.FromSeconds(5));
});

httpClientBuilder.AddStandardHedgingHandler();

httpClientBuilder.AddStandardHedgingHandler(static (IRoutingStrategyBuilder builder) =>
{
    // Hedging allows sending multiple concurrent requests
    builder.ConfigureOrderedGroups(static orderedGroup =>
    {
        orderedGroup.Groups.Add(new EndpointGroup()
        {
            Endpoints =
            {
                // Imagine a/b testing
                new WeightedEndpoint() { Uri = new("https://example.net/api/a"), Weight = 01 },
                new WeightedEndpoint() { Uri = new("https://example.net/api/b"), Weight = 99 }
            }
        });

        // Add more groups, having only one means that hedging will never 
        // do secondary attempts. The maximum number of attempts directly 
        // correlates to the number of groups.
    });

    builder.ConfigureWeightedGroups(static weightedGroup =>
    {
        weightedGroup.SelectionMode = WeightedGroupSelectionMode.EveryAttempt;

        weightedGroup.Groups.Add(new WeightedEndpointGroup()
        {
            Endpoints =
            {
                // imagine a/b testing
                new WeightedEndpoint() { Uri = new("https://example.net/api/light"), Weight = 01 },
                new WeightedEndpoint() { Uri = new("https://example.net/api/heavy"), Weight = 99 }
            }
        });

        // Add more groups, having only one means that hedging will never 
        // do secondary attempts. The maximum number of attempts directly 
        // correlates to the number of groups.
    });
});

// Bind the options to a strongly-typed object
HttpRetryStrategyOptions options = new();

// Get the retry options from the configuration
builder.Configuration
    .GetSection("RetryOptions")
    .Bind(options);

httpClientBuilder.AddResilienceHandler(
    "AdvancedPipeline",
    static (ResiliencePipelineBuilder<HttpResponseMessage> builder,
        ResilienceHandlerContext context) =>
{
    // Enable the reloads whenever the named options change
    context.EnableReloads<RetryStrategyOptions>("RetryOptions");

    // Retrieve the named options
    var retryOptions =
        context.ServiceProvider
            .GetRequiredService<IOptionsMonitor<RetryStrategyOptions<HttpResponseMessage>>>()
            .Get("RetryOptions");

    // Add retries using the resolved options
    builder.AddRetry(retryOptions);
});

var app = builder.Build();

var client = app.Services.GetRequiredService<ExampleClient>();

await foreach (var comment in client.GetCommentsAsync())
{
    Console.WriteLine(comment);
}

app.Run();
```

Here's the `ExampleClient` implementation:

```cs
using System.Net.Http.Json;

internal sealed class ExampleClient(HttpClient client)
{
    public IAsyncEnumerable<Comment?> GetCommentsAsync()
    {
        return client.GetFromJsonAsAsyncEnumerable<Comment>("/comments");
    }
}
```

And the _Comment.cs_ file:

```cs
public record class Comment(
    int PostId, int Id, string Name, string Email, string Body);
```

### Expected behavior

I'd expect this to run without throwing.

### Actual behavior

It throws:

> **System.InvalidOperationException**: No service for type 'Microsoft.Extensions.Compliance.Redaction.IRedactorProvider' has been registered.'

### Known Workarounds

Maybe adding a `PackageReference` to the compliance/redaction bits, but I'm not sure. And I don't have time to test that. ☹️

### Configuration

<details>
<summary>dotnet --info</summary>

.NET SDK:
 Version:           8.0.100-rtm.23512.16
 Commit:            cbfd87357c
 Workload version:  8.0.100-manifests.f8016506

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.100-rtm.23512.16\

.NET workloads installed:
 Workload version: 8.0.100-manifests.f8016506
 [maui-windows]
   Installation Source: VS 17.7.34031.279
   Manifest Version:    8.0.0-rc.2.9373/8.0.100-rc.2
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.2\microsoft.net.sdk.maui\8.0.0-rc.2.9373\WorkloadManifest.json
   Install Type:              Msi

 [maui-maccatalyst]
   Installation Source: VS 17.7.34031.279
   Manifest Version:    8.0.0-rc.2.9373/8.0.100-rc.2
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.2\microsoft.net.sdk.maui\8.0.0-rc.2.9373\WorkloadManifest.json
   Install Type:              Msi

 [maccatalyst]
   Installation Source: VS 17.7.34031.279
   Manifest Version:    17.0.8314-ci.release-test-net8-0-xcode15-multi-targeting/8.0.100-rc.2
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.2\microsoft.net.sdk.maccatalyst\17.0.8314-ci.release-test-net8-0-xcode15-multi-targeting\WorkloadManifest.json
   Install Type:              Msi

 [maui-ios]
   Installation Source: VS 17.7.34031.279
   Manifest Version:    8.0.0-rc.2.9373/8.0.100-rc.2
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.2\microsoft.net.sdk.maui\8.0.0-rc.2.9373\WorkloadManifest.json
   Install Type:              Msi

 [ios]
   Installation Source: VS 17.7.34031.279
   Manifest Version:    17.0.8314-ci.release-test-net8-0-xcode15-multi-targeting/8.0.100-rc.2
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.2\microsoft.net.sdk.ios\17.0.8314-ci.release-test-net8-0-xcode15-multi-targeting\WorkloadManifest.json
   Install Type:              Msi

 [maui-android]
   Installation Source: VS 17.7.34031.279
   Manifest Version:    8.0.0-rc.2.9373/8.0.100-rc.2
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.2\microsoft.net.sdk.maui\8.0.0-rc.2.9373\WorkloadManifest.json
   Install Type:              Msi

 [android]
   Installation Source: VS 17.7.34031.279
   Manifest Version:    34.0.3/8.0.100-rc.2
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.2\microsoft.net.sdk.android\34.0.3\WorkloadManifest.json
   Install Type:              Msi

 [wasm-tools]
   Installation Source: VS 17.7.34031.279
   Manifest Version:    8.0.0-rtm.23511.16/8.0.100-rtm
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rtm\microsoft.net.workload.mono.toolchain.current\8.0.0-rtm.23511.16\WorkloadManifest.json
   Install Type:              Msi

 [aspire]
   Installation Source: VS 17.9.34215.31
   Manifest Version:    8.0.0-preview.1.23505.4/8.0.100-rc.1
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100-rc.1\microsoft.net.sdk.aspire\8.0.0-preview.1.23505.4\WorkloadManifest.json
   Install Type:              Msi

Host:
  Version:      8.0.0-rtm.23511.16
  Architecture: x64
  Commit:       256bf22a3d

.NET SDKs installed:
  6.0.318 [C:\Program Files\dotnet\sdk]
  7.0.112 [C:\Program Files\dotnet\sdk]
  7.0.401 [C:\Program Files\dotnet\sdk]
  8.0.100-rc.2.23502.2 [C:\Program Files\dotnet\sdk]
  8.0.100-rtm.23512.16 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.22 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.23 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.11 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0-rc.2.23480.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0-rtm.23512.10 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.22 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.23 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0-rc.2.23479.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0-rtm.23511.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.22 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.23 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.11 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.0-rc.2.23479.10 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.0-rtm.23512.4 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
</details>

Tagging @joperezr

**Comment by martintmk** [comment]
@IEvangelist The latest resilience code does not depend on redaction anymore. This won't be an issue when the next version is released.

See #4470

**Comment by joperezr** [comment]
@IEvangelist can you try with version `8.0.0-rtm.23517.2` to validate you don't hit the issue any longer? You will need the dotnet8 feed for that: `https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8/nuget/v3/index.json`

**Comment by IEvangelist** [comment]
Yes, this is fixed -- thank you!

**Comment by IEvangelist** [comment]
Fixed by #4473
